### DISCUSSION TITLE
I get recreate error when I use redirect in approuter

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#71560] [App Router]
I get recreate error when I use redirect in approuter

```
[Error] Server Error: NEXT_REDIRECT

The above error occurred in the component.

React will try to recreate this component tree from scratch using the error boundary you provided, RedirectErrorBoundary.

[Error] Error: Rendered more hooks than during the previous render.

The above error occurred in the component.

React will try to recreate this component tree from scratch using the error boundary you provided, ErrorBoundaryHandler.
```

**Comment by icyJoseph** [comment]
Show us how you are using it.

**Reply by elie222** [reply]
Started happening with the Next.js 15 update.

Example:

```tsx
import { redirect } from "next/navigation";
import { isPremium } from "@/utils/premium";
import { auth } from "@/app/api/auth/[...nextauth]/auth";
import prisma from "@/utils/prisma";
import { env } from "@/env";

export async function checkAndRedirectForUpgrade() {
  if (!env.NEXT_PUBLIC_WELCOME_UPGRADE_ENABLED) return;

  const session = await auth();

  const email = session?.user.email;

  if (!email) redirect("/login");

  const user = await prisma.user.findUnique({
    where: { email },
    select: {
      premium: { select: { lemonSqueezyRenewsAt: true } },
      completedAppOnboardingAt: true,
    },
  });

  if (!user) redirect("/login");

  if (!isPremium(user.premium?.lemonSqueezyRenewsAt || null)) {
    if (!user.completedAppOnboardingAt) redirect("/onboarding");
    else redirect("/welcome-upgrade");
  }
}
```

And I saw the error when this was called:

```tsx
redirect("/onboarding");
```

**Reply by and0m** [reply]
Getting the same error. I have a modal that gets rendered under certain conditions.

Here is the code for my page
```
export default async function Home({ searchParams }: { searchParams: { [key: string]: string | string[] | undefined } })
{
  const condition: boolean = await checkCondition();
  const recParams: any = await searchParams;
  if (condition && recParams.modal === undefined)
  {
    const modalData: string = ModalMessages.get();
    return redirect("/" + modalData);
  }
return (
<>
<Modal />
<ComponentA />
</>
)
}
```

If the condition is true and I am not rendering the modal then redirect, otherwise just render the page. The Modal component is hidden by default and shows up if the redirect has certain parameters.

This started happening after upgrading to Next 15.

**Reply by icyJoseph** [reply]
Do you think, you could try to setup a reproduction repository?

**Comment by ghost** [comment]
> Show us how you are using it.

redirect("/example")

**Comment by jacobcwright** [comment]
upgraded to next 15.0.3 and it was resolved

**Reply by FurkanTokmak** [reply]
I am using 15.1.0 and I have the issue.

**Comment by pa4080** [comment]
I was experiencing the same problem with my App and Next.js 15 - when `redirect('/somewhere')` function is called from **if** statement. I think with Next.js 14 the problem wasn't presented...

I've spent some time investigating and discovered the following hydration error workaround was causing the problem in my case.

```tsx
"use client";

import { ThemeProvider as NextThemesProvider } from "next-themes";
import { type ThemeProviderProps } from "next-themes/dist/types";
import { useEffect, useState } from "react";

export function ThemeProvider({ children, ...props }: ThemeProviderProps) {
  const [mounted, setMounted] = useState(false);
  useEffect(() => { setMounted(true) }, []);
  
  return mounted && <NextThemesProvider {...props}>{children}</NextThemesProvider>;
}
```

So in order to **solve the problem** I've refactored the application in this way:

- `app/layout.tsx`
```tsx
import { Suspense } from "react";
import "globals.css";
import ThemeProviderClient from "./_theme/ThemeProviderClient";

export default function RootLayout({ children }: Readonly<{ children: React.ReactNode }>) {
  return (
    <html lang="en">
      <body>
        <ThemeProviderClient>{children}</ThemeProviderClient>
      </body>
    </html>
  );
}
```

- `app/_theme/ThemeProviderClient.tsx`
```tsx
"use client";

import dynamic from "next/dynamic";

const ThemeProvider = dynamic(() => import("./ThemeProvider").then((mod) => mod.ThemeProvider), {
  ssr: false,
  loading: () => <div className="flex items-center justify-center w-screen h-screen">Loading...</div>,
});

export default function ThemeProviderClient({ children }: Readonly<{ children: React.ReactNode }>) {
  return (
    <ThemeProvider disableTransitionOnChange enableSystem attribute="class" defaultTheme="system">
      {children}
    </ThemeProvider>
  );
}
```

- `app/_theme/ThemeProvider.tsx`
```tsx
import { ThemeProvider as NextThemesProvider } from "next-themes";
import { type ThemeProviderProps } from "next-themes/dist/types";

export function ThemeProvider({ children, ...props }: ThemeProviderProps) {
  return <NextThemesProvider {...props}>{children}</NextThemesProvider>;
}
```

I think all this magic should happen under the hoot within `next-themes/ThemeProvider`, but it's not working yet.

**Comment by pa4080** [comment]
Unfortunately the refactoring shown in the above comment solve the issue at one certain place, but the problem exist at all. 

When I do redirect from a "server action", and the fiction is called outside a try block, the redirection went well.

It looks this is related with this topic: [Next 13.4 Error: NEXT_REDIRECT in API routes](https://stackoverflow.com/questions/76191324/next-13-4-error-next-redirect-in-api-routes)

**Reply by geekysaurabh001** [reply]
> Unfortunately the refactoring shown in the above comment solve the issue at one certain place, but the problem exist at all.
> 
> When I do redirect from a "server action", and the fiction is called outside a try block, the redirection went well.
> 
> It looks this is related with this topic: [Next 13.4 Error: NEXT_REDIRECT in API routes](https://stackoverflow.com/questions/76191324/next-13-4-error-next-redirect-in-api-routes)

Did you find the solution?
